src/mono/wasm/runtime/net6-legacy/hybrid-globalization.ts

```diff
@@ -84,6 +84,129 @@ export function pass_exception_details(ex: any, exceptionMessage: Int32Ptr){
+    {
+        return;
+    }
+    // pop is O(1) while shift is O(n) and reverse is O(n)
+    // it pays off to revert twice and pop more than shift multiple times
+    // unless we have < 2 empty chars at the beginning
+    if (segment.length === 1 || (segment.length > 1 && "".localeCompare(segment[1].segment, undefined) !== 0))
+    {
+        segment.shift();
+        return;
+    }
+    segment.reverse();
+    skip_zeroWidthChars_from_end(segment);
+    segment.reverse();
+}
+
+function skip_zeroWidthChars_from_end(segment: Intl.SegmentData[])
+{
+    while (segment.length !== 0 && "".localeCompare(segment[segment.length - 1].segment, undefined) === 0)
+    {
+        segment.pop();
+    }
+}
+
+export function mono_wasm_ends_with(exceptionMessage: Int32Ptr, culture: MonoStringRef, str1: number, str1Length: number, str2: number, str2Length: number, options: number, matchLengthPointer: number): number{
+    const cultureRoot = mono_wasm_new_external_root<MonoString>(culture);
+    try{
+        const cultureName = conv_string_root(cultureRoot);
+        const suffix = get_utf16_string(str2, str2Length);
+        // no need to look for an empty string
+        const result = "".localeCompare(suffix, undefined);
+        if (result === 0)
+            return 1; // true
+
+        const source = get_utf16_string(str1, str1Length);
+        const locale = cultureName ? cultureName : undefined;
+        const graphemesSource = segment_string_locale_sensitive(source, locale);
+        const graphemesSuffix = segment_string_locale_sensitive(suffix, locale);
+
+        // zero Width chars at the end do not matter in string comparison but are useful for calculating the match length
+        skip_zeroWidthChars_from_end(graphemesSource);
+        // zero Width chars do not matter in prefix's both ends
+        skip_zeroWidthChars_from_start(graphemesSuffix);
+        skip_zeroWidthChars_from_end(graphemesSuffix);
+
+        const casePicker = (options & 0x1f);
+        for (let i = 0; i < graphemesSuffix.length; i++)
+        {
+            const isEqual = compare_strings(graphemesSource[graphemesSource.length - 1 - i].segment, graphemesSuffix[graphemesSuffix.length - 1 - i].segment, locale, casePicker);
+            if (isEqual !== 0)
+                return 0; // false
+        }
+        const offset = graphemesSource.length - graphemesSuffix.length;
+        const firstCharIdx = graphemesSource[offset].index;
+        const matchLen = str1Length - firstCharIdx;
+        setU32(matchLengthPointer, matchLen);
+        return 1; // true
+    }
+    catch (ex: any) {
+        pass_exception_details(ex, exceptionMessage);
+        return -1;
+    }
+    finally {
+        cultureRoot.release();
+    }
+}
+
+export function segment_string_locale_sensitive(string: string, locale: string | undefined) : Intl.SegmentData[]
+{
+    const segmenter = new Intl.Segmenter(locale, { granularity: "grapheme" });
```

Comment:
I would appreciate comment about why we are segmenting the string into  "user perceived letters" instead of comparing the whole string ? Is that because of the need to know the match index ?

Would it be possible it to assume that most calls return false (no match) and optimize for it by comparing unsegmented string first ?

Is creation of `Segmenter` object expensive ?

Reply:
It's because we need locale-sesitive comparison. We cannot guess how many chars does n letters consist of. Consider the case:
```
s1 = "A\u0300" // 2 chars, 1 letter
s2 = "\u00C0" // 1 char, 1 letter
```
segmenter recognises both as 1 letter and we can easily compare them. If we were to compare s1[0] === s2[0] then we would get false.

Reply:
Could we normalize both strings instead ?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize